Psa import export

ronald-cron-arm opened the pull request:
## Description
Rework PSA key import and export as described in the documentation file docs/architecture/psa-crypto-implementation-structure.md added in this PR.
This work was primarily triggered by testing considerations. In terms of testing of the driver interface and its integration into the
PSA application interface implementation, with this PR:

- all key import/export cryptographic operations are triggered by a call to the import/export driver interface. The call is the same for hardware and software implementations. The import/export driver interface is thus exercised by all the PSA unit tests that import/export a key as opposed to just the driver wrapper tests before this PR.
- we have now three transparent test drivers to play with: the ECP one, the RSA one and the one combining the two.
- the code of the transparent test driver is much simpler.
- the code of the transparent test driver do not rely directly on the Mbed TLS APIs but on PSA software drivers. This is good to have with the coming switch to the PSA interface as the main application interface of the project. 
- the all.sh test_psa_crypto_drivers component runs all PSA unit tests with all ECC/RSA import and export cryptographic operations handled by the transparent test driver.
- the all.sh test_psa_crypto_config_basic component runs all PSA unit tests with all ECP import and export cryptographic operations handled by the transparent test driver with the software implementations (`psa_crypto_ecp_import_key` and `psa_crypto_ecp_export_public_key`) not included in the Mbed TLS library.

## Status
READY

## Requires Backporting
NO, PSA only

gilles-peskine-arm reviewed (commented):
I looked at the general approach. This is not a complete review (I left a couple of comments about specific points, but only because I happened to notice them).

I don't think this meets the objective of testing support for drivers when the core doesn't include the corresponding mechanism. I would have expected that, for example, we could build `libmbedcryto.o` that doesn't have any support for RSA, and then the tests would link things like `mbedtls_transparent_test_driver_rsa_import_key` and `rsa.o` separately.

ronald-cron-arm commented on library/psa_crypto_rsa.h:
```diff
@@ -0,0 +1,136 @@
+ *                          contents of the context and the context itself
+ *                          when done.
+ */
+psa_status_t psa_crypto_rsa_load_representation( psa_key_type_t type,
```
I can change this if you are convinced by an alternative. Not having `psa` in the function name open the door to name conflicts with the MbedTLS classic functions but that's maybe just a theoretical concern.

gilles-peskine-arm replied:
We control the `mbedtls_` namespace, more so than the `psa_` namespace (even `psa_crypto_`. We can use `mbedtls_psa_` (which is informally growing for things that bridge the two APIs in either direction, as well as for pure-PSA things that are specific to Mbed TLS, but it isn't used systematically).

ronald-cron-arm replied:
ok thus the name of the file would be mbedtls_psa_rsa.c and the functions prefixed by mbedtls_psa_ ?

gilles-peskine-arm replied:
The file name should stay `psa_xxx`. We have `rsa.c`, not `mbedtls_rsa.c`.

I currently favor function names like `mbedtls_psa_rsa_load_representation`. It would be good to discuss long-term plans at some point and establish clear naming conventions that are compatible with these long-term plans. But establishing these conventions is a longer-term thing than the driver testing work and should not hold it back.

ronald-cron-arm replied:
ok thus `psa_crypto_rsa.c` for the file name and `mbedtls_psa_rsa_xyz` for the function names, right?

gilles-peskine-arm replied:
sounds good

ronald-cron-arm replied:
ok cool, I will go for that then, thanks.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1010,122 +557,76 @@ psa_status_t psa_copy_key_material_into_slot( psa_key_slot_t *slot,
-static psa_status_t psa_import_key_into_slot( psa_key_slot_t *slot,
-                                              const uint8_t *data,
-                                              size_t data_length )
+psa_status_t psa_import_key_into_slot(
```
It's the equivalent of the "_internal()" functions for the import operation. I kept the original name because "_into_slot" compared to "_internal" brings some useful information to me. The function is called from the driver wrapper to dispatch depending on the type of the key. I agree that eventually the dispatch should be in the driver wrapper but that's additional changes, more commits and it is already quite big.

Otherwise, the documentation is not gone, it is in psa_crypto_core.h.

gilles-peskine-arm replied:
Existence: ok then, makes sense for it to be future work.

Documentation: ah, right, makes perfect sense.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -2753,7 +2753,7 @@ depends_on:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_ECDSA_C
 
 PSA generate key: ECC, Curve25519, good
-depends_on:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_CURVE25519_ENABLED:MBEDTLS_ECDH_C
+depends_on:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_CURVE25519_ENABLED:MBEDTLS_ECDH_C:!MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR
```
This is temporary and removed in the PR related to generate_key. I've outlined this in the commit message.

gilles-peskine-arm replied:
Ok (this wasn't fundamental and I did a time-bound review so I didn't take the time to investigate why or read the commit message).

ronald-cron-arm commented on library/psa_crypto_rsa.c:
```diff
@@ -0,0 +1,318 @@
+ * BEYOND THIS POINT, TEST DRIVER ENTRY POINTS ONLY.
+ */
+
+#if defined(PSA_CRYPTO_DRIVER_TEST)
```
We discussed about that offline. This wasn't in my plans but you convinced me that when I am done with the rework of the driver entry points (import/export/generate/sign/verify/cipher) this is the top priority for driver interface testing improvement.

ronald-cron-arm commented on docs/architecture/psa-crypto.md:
```diff
@@ -0,0 +1,50 @@
+PSA Cryptograpy API implementation and PSA driver interface
```
Yes I know this is a bad name but couldn't figure out a good one: psa-crypto-implementation.md, psa_crypto-implementation-and-unified-driver-interface.md ?

gilles-peskine-arm replied:
`psa-crypto-dispatch.md`? `psa-crypto-implementation-structure.md`?

ronald-cron-arm replied:
thanks for the proposals, psa-crypto-implementation-structure.md looks quite good to me, I will go for that.

ronald-cron-arm commented:
I've done the renaming and rebased on top of the 2.25 release. If necessary the previous version can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-import-export-1) .

ronald-cron-arm commented:
I've just added the "needs reviewer" label as Gilles has done a high-level design review of the PR but it is unlikely that he will find time to do a complete review.

mpg reviewed (changes requested):
I'm mostly happy with this PR. I spent a lot of time on it and in the end only have pretty minor comments.

Thanks for adding a document about the implementation structure of the driver interface, I think in general we should have more of those. Actually, I think this one could probably be expanded in two directions:

1. More details about the implementation: where allocation/freeing happens and through which functions, naming conventions, what should like in what file, the place of the default software implementation in the dispatch stack, etc.
2. The testing strategy, which may or may not be a separate document.

I'm not necessarily asking for this to happen in this PR, as I realise things are very much in flux right now, especially on the testing front, but I just wanted to point out that I think that's something we should do at some point, preferably when things are still fresh. That would be good both for us and for external contributors.

mpg commented:
Note: out of curiosity I've check the impact on code size (as the code gets more uniform, as the PR description says, I wondered if that meant smaller code size). Turns out it's a very small increase in code size, which I think doesn't matter.

Tested with:
```
scripts/config.py baremetal
make clean
make CC=arm-none-eabi-gcc CFLAGS='-Os -march=armv6-m' lib >/dev/null
arm-none-eabi-size -t library/*.a | sed -n '1p;$p'
```
Before: 428050.
After: 428348.
Diff: +298 bytes - not a problem.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1476,58 +1476,114 @@ static psa_status_t psa_export_key_buffer_internal( const psa_key_slot_t *slot,
+    {
+        if( ( drv->key_management == NULL   ) ||
+            ( drv->key_management->p_export == NULL ) )
+            return( PSA_ERROR_NOT_SUPPORTED );
```
Fix in psa_crypto_driver_wrappers.c as this code has been moved to this file.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1476,58 +1476,114 @@ static psa_status_t psa_export_key_buffer_internal( const psa_key_slot_t *slot,
      * pointer and we can do things like memset(data, ..., data_size). */
     if( data_size == 0 )
-        return( PSA_ERROR_BUFFER_TOO_SMALL );
+    {
```
That's definitely better.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1569,24 +1447,12 @@ psa_status_t psa_export_public_key_internal(
-            mbedtls_free( rsa );
-
-            return( status );
+            return( mbedtls_psa_rsa_export_public_key( attributes,
```
Roger that.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -731,7 +719,15 @@ static psa_status_t psa_import_key_into_slot( psa_key_slot_t *slot,
+                                         data, data_length,
+                                         slot->key.data, data_length,
+                                         &slot->key.bytes );
+            return( status );
```
Yes, if we return in error here, the key creation fails and eventually the psa_fail_key_creation() will be called. The buffer is de-allocated as part of the call to psa_fail_key_creation().

mpg replied:
Thanks!

ronald-cron-arm commented on docs/architecture/psa-crypto.md:
````diff
@@ -0,0 +1,50 @@
+     * clean-up in case of error ... */
+}
+```
+Obviously, for some PSA APIs the implementation is more complicated with several potential conditionnal calls to different driver entry points but most of PSA API code aimed to be organized along those lines. The implementation of the `psa_driver_wrapper_<entry_point>` function is generated by the build system based on the JSON driver description files of the various PSA drivers making up the Mbe ...
````
I've added a note about that.

stevew817 commented on library/psa_crypto.c:
```diff
@@ -2302,39 +1786,34 @@ psa_status_t psa_import_key( const psa_key_attributes_t *attributes,
+        goto exit;
     }
+
     status = psa_validate_optional_attributes( slot, attributes );
```
The call to `validate_optional_attributes` here will still rely on having a software implementation compiled in for RSA keys. That part of the validation would/should have been done as part of the `wrapper_import_key` call chain, such that the call tree after returning from a driver wrapper call does not rely on software libraries. Returning 'success' from the `wrapper_import_key` call evidently means the key is deemed valid.

ronald-cron-arm replied:
Yes, that makes sense to me. The RSA code checking the domain parameters in psa_validate_optional_attributes() should probably be moved to rsa_import_key() in psa_crypto_rsa.c. @gilles-peskine-arm, what do you thing about this?

gilles-peskine-arm replied:
The validation of domain parameters also needs to be done when generating or deriving a key. Embedding it in import is definitely not right.

`psa_generate_key` should call `psa_validate_optional_attributes` (or maybe a refactored version of it). The reason it doesn't is that no use of domain parameters has been implemented yet in `psa_generate_key`, and the other validations are redundant for key generation, so it gets away with just requiring the domain parameters to be `NULL`. Which is a pity because the most asked feature that domain parameters would solve is generating an RSA key with e=3, and we're almost there.

Making it even harder for `psa_generate_key` to use domain parameters is not useful: if domain parameters are really blocking driver support, we should remove domain parameters. I hope we don't have to go that far.

Why can't we reserve domain parameters for the software case, and worry later about adding domain parameter support to the driver interface? If there's a driver, just require the domain parameters to be null.

----

The current state of domain parameters in Mbed TLS is unsatisfactory: there's some infrastructure, but not enough to do anything useful. Domain parameters were in a draft of the PSA Crypto API specification, but were removed from the 1.0 release because the experience of implementing it in Mbed TLS showed that the current form is a hassle. We haven't removed domain parameters from Mbed TLS because we do want the features they bring, but work on them is paused while we focus on other things such as driver support.

We've identified three uses for domain parameters in the currently declared algorithms:

* Specifying a non-default public exponent in RSA key generation. RSA is implemented in Mbed TLS, but a custom public exponent is not possible through the PSA API (it is possible through the classic API).
* Using a custom group for finite-field Diffie-Hellman. FFDH is implemented in Mbed TLS, but not yet through the PSA API, and even if it was a first step would be to only support predefined groups.
* Using DSA (the non-EC kind). I doubt that DSA will never be implemented in Mbed TLS.

Because the domain parameter interface is likely to change, I want to minimize its entanglement in the driver interface.

ronald-cron-arm replied:
From crypto_extra.h:
```
- For RSA keys (#PSA_KEY_TYPE_RSA_PUBLIC_KEY or #PSA_KEY_TYPE_RSA_KEY_PAIR), 
 *   the domain parameter data consists of the public exponent,                 
 *   represented as a big-endian integer with no leading zeros.                 
 *   This information is used when generating an RSA key pair.                  
 *   **When importing a key, the public exponent is read from the imported        
 *   key data and the exponent recorded in the attribute structure is ignored.**  
 *   As an exception, the public exponent 65537 is represented by an empty      
 *   byte string.  
 ```
Thus should we just remove the check of the RSA exponent when importing a key ? That sounds sensible to me.

gilles-peskine-arm replied:
This text was added in https://github.com/ARMmbed/mbedtls-psa/pull/263 (private link). Looking back, it described the implementation at the time (import didn't check the RSA exponent), but I think this was an implementation limitation (domain parameters for RSA were a new thing) and not a desirable specification. Having a function ignore an explicit input is not a good design. https://github.com/ARMmbed/mbed-crypto/pull/101 added the check, which I think was the right thing, but forgot to update the documentation.

ronald-cron-arm replied:
Thanks for not blocking the PR on this. I agree that there is an issue and I've described it [here](https://github.com/ARMmbed/mbedtls/issues/4095). I am not addressing it here as I don't see a way of doing it in one or two not controversial commits.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
````diff
@@ -0,0 +1,73 @@
+```
+Obviously, for some PSA APIs the implementation is more complicated with several potential conditional calls to different driver entry points but most of PSA API code aims to be organized along those lines.
+
+The implementations of `psa_driver_wrapper_<entry_point>` functions are generated by the build system based on the JSON driver description files of the various PSA drivers making up the Mbed TLS PSA Cryptography API implementation. The implementations are generated in a psa_crypto_driver_wrappers_generated.c C file and the function prototypes declared in a psa_crypto_driver_wrappers_generated.h h ...
````
Yes, I thought it would be better to emphasize in the name that the file are generated but that's not their current name. I am changing to psa_crypto_driver_wrappers.[ch] to not confuse everybody. If we decide to change the name of those files at some point we will update them here as well.

ronald-cron-arm commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -410,85 +409,208 @@ psa_status_t psa_driver_wrapper_generate_key( const psa_key_attributes_t *attrib
 #endif /* PSA_CRYPTO_DRIVER_TEST */
+#endif /* PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT */
+            /* Fell through, meaning no accelerator supports this operation */
+            return( psa_import_key_into_slot( attributes,
```
Yes that's an exception. I've done it like that because this name brings to me some useful extra information compared to just psa_import_key_internal(): the key is imported into a key slot. But I agree that this is a breach of consistency. I can change it if you prefer. On this subject, it is also worth mentioning that all _internal() functions are aimed to disappear when psa_crypto_driver_wrappers.c will be generated. Their code will be replaced by generated code.

ronald-cron-arm commented on library/psa_crypto_core.h:
```diff
@@ -208,4 +214,97 @@ psa_status_t psa_copy_key_material_into_slot( psa_key_slot_t *slot,
+ */
+mbedtls_ctr_drbg_context *psa_get_ctr_drbg_context( void );
+
+/** Import a key in binary format.
```
The prototype of the driver entry points are in separate files:  psa_crypto_rsa.h and psa_crypto_ecp.h. The _internal() functions that dispatch (not only, some more refactoring is needed here) to driver entry points are declared in psa_crypto_core.h. As mentioned in the previous comment those functions are planned to disappear when psa_crypto_driver_wrappers.c is generated. We could get rid of them yet but it would be just more commits and we don't really need that for the time being.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
````diff
@@ -0,0 +1,73 @@
+     * clean-up in case of error ... */
+}
+```
+Obviously, for some PSA APIs the implementation is more complicated with several potential conditional calls to different driver entry points but most of PSA API code aims to be organized along those lines.
````
I don't have anything very specific in mind actually but it is likely that we won't be able to do an one to one mapping between the PSA APIs and the driver entry points. That's what I try to say here. I've rephrased it. Hopefully this is clearer.

stevew817 replied:
I think it's still a bit of two messy paragraphs. You're repeating the point about 'several calls to the driver interface' twice. I also think the word 'obviously' shouldn't be in any documentation, ever. If it were that obvious, you wouldn't write about it.

mpg replied:
I agree that there is redundancy between the two paragraphs and it does not make things clearer. I'd suggest replacing the second paragraph with "For example, [description of an example, perhaps MAC verify]. But except for that kind of case, most of PSA API code should be organized as above."

ronald-cron-arm replied:
The first version of the PR contained only the second paragraph. Then, I intended to replace this paragraph with what is not the first paragraph of the two but forgot the remove the initial version. Thus yes there is redundancy as the second paragraph shouldn't be there.  I am removing it and trying to further improve with an example.

ronald-cron-arm commented:
@stevew817 you have posted comments on this PR and I think I have addressed them. Are you ok to be the second reviewer for this PR?

stevew817 commented on library/psa_crypto_core.h:
```diff
@@ -208,4 +214,97 @@ psa_status_t psa_copy_key_material_into_slot( psa_key_slot_t *slot,
+ * \param[in]  data             The buffer containing the key data in import
+ *                              format.
+ * \param[in]  data_length      Size of the \p data buffer in bytes.
+ * \param[out] key_buffer       The buffer containing the key data in output
```
Minor: would be clearer stating "The buffer to contain the key data in output format upon successful return" since the buffer isn't supposed to contain anything on function entry.

I know it's implied by marking it as an output, but still.

stevew817 reviewed (dismissed):
I only have very non-blocking comments outstanding.

I haven't checked out and compiled this code myself, but the CI seemed happy before the merge conflict.

mpg reviewed (approved):
Thanks for addressing my feedback and expanding the documentation. I'm happy with the result.

mpg commented:
I quickly went over Gilles's feedback and made sure it had been addressed, resolving the conversations in order to reflect that.

Now I think the only thing preventing this PR from getting merged is the conflicts. @ronald-cron-arm can you resolve them and ping us once you think the PR is ready for a (hopefully) final review?

mpg reviewed (dismissed):
Still looks good to me. Conflicts are blocking this though.

ronald-cron-arm commented:
@stevew817, before to rebase, I've addressed your 3 last comments by reworking the now penultimate commit ("Expand and improve psa-crypto-implementation-structure.md"), adding a commit ("psa: Improve psa_import_key_into_slot() documentation") and creating issue #4095.
The result of this can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-import-export-2) and the diff compared to the previous version [here](https://github.com/ARMmbed/mbedtls/compare/f7d10e22af5e1420d6b96b7c7372e16745fa15e0..6f677124b8d44c7d9fa572436da77db5f317363b).
Then I've rebased on top of c539b95.
Here are the non-trivial conflicts I've been through:

===============================================================================

Commit "psa: Make CTR DRBG available to other PSA C modules"

Conflict with 5894e8e7a4 that also "exported" the CTR DRBG context but
as part of a larger re-organization of the PSA RNG code.

Resolution:skipped the commit

===============================================================================

Commit "psa: se: Create key context in SE key slots"

In psa_crypto.c, conflict with 462aa575a4 (PR 3969)

<<<<<<< HEAD                                                                    
#if defined(MBEDTLS_PSA_CRYPTO_SE_C)                                            
    if( psa_get_se_driver( slot->attr.lifetime, NULL, NULL ) &&                 
        psa_key_slot_is_external( slot ) )                                      
    {                                                                           
        /* No key material to clean. */                                         
    }                                                                           
    else                                                                        
#endif /* MBEDTLS_PSA_CRYPTO_SE_C */                                            
    {                                                                           
        /* Data pointer will always be either a valid pointer or NULL in an     
         * initialized slot, so we can just free it. */                         
        if( slot->data.key.data != NULL )                                       
            mbedtls_platform_zeroize( slot->data.key.data, slot->data.key.bytes);
        mbedtls_free( slot->data.key.data );                                    
        slot->data.key.data = NULL;                                             
        slot->data.key.bytes = 0;                                               
    }                                                                           
=======                                                                         
    /* Data pointer will always be either a valid pointer or NULL in an         
     * initialized slot, so we can just free it. */                             
    if( slot->key.data != NULL )                                                
        mbedtls_platform_zeroize( slot->key.data, slot->key.bytes);             
                                                                                
    mbedtls_free( slot->key.data );                                             
    slot->key.data = NULL;                                                      
    slot->key.bytes = 0;                                                        
>>>>>>> psa: se: Create key context in SE key slots

Resolution: kept "psa: se: Create key context in SE key slots" code only as
the code introduced by 462aa575a4 is not needed anymore.

-------------------------------------------------------------------------------

In psa_crypto_slot_management.c, conflict with ac3434fc19 (PR 3996)

<<<<<<< HEAD                                                                    
        memcpy( &slot->data.se.slot_number, &data->slot_number,                 
                sizeof( slot->data.se.slot_number ) );                          
                                                                                
        status = PSA_SUCCESS;                                                   
        goto exit;                                                              
=======                                                                         
        status = psa_copy_key_material_into_slot(                               
                     slot, data->slot_number, sizeof( data->slot_number ) );    
>>>>>>> psa: se: Create key context in SE key slots

Resolution: Don't forget the "goto exit;" added by ac3434fc19 to avoid calling
            twice psa_copy_key_material_into_slot().

        status = psa_copy_key_material_into_slot(                               
                     slot, data->slot_number, sizeof( data->slot_number ) );    
        goto exit;

===============================================================================

Commit: "psa: Move RSA/ECP export code to RSA/ECP specific C file"

In psa_crypto.c, conflict with 5894e8e7a4 and the usage of
MBEDTLS_PSA_RANDOM_STATE.

Resolution: In psa_crypto.c, nothing. Just removed mbedtls_psa_ecp_export_key().
            In psa_crypto_ecp.c:mbedtls_psa_ecp_export_key(), replaced

            status = mbedtls_to_psa_error(
                mbedtls_ecp_mul( &ecp->grp, &ecp->Q, &ecp->d, &ecp->grp.G,
                                 mbedtls_ctr_drbg_random,
                                 psa_get_ctr_drbg_context( ) ) );
            by

            status = mbedtls_to_psa_error(
                mbedtls_ecp_mul( &ecp->grp, &ecp->Q, &ecp->d, &ecp->grp.G,
                                 mbedtls_psa_get_random,
                                 MBEDTLS_PSA_RANDOM_STATE ) );

===============================================================================

Commit: "psa: import: Move registered SE support to the driver wrapper"

In psa_crypto.c, conflict with ac3434fc19 (PR 3996)

<<<<<<< HEAD                                                                    
    else                                                                        
#endif /* MBEDTLS_PSA_CRYPTO_SE_C */                                            
    if( psa_key_lifetime_is_external( psa_get_key_lifetime( attributes ) ) )    
    {                                                                           
        /* Importing a key with external lifetime through the driver wrapper    
         * interface is not yet supported. Return as if this was an invalid     
         * lifetime. */                                                         
        status = PSA_ERROR_INVALID_ARGUMENT;                                    
        goto exit;                                                              
    }                                                                           
    else                                                                        
    {                                                                           
       /* In the case of a transparent key or an opaque key stored in local     
        * storage (thus not in the case of generating a key in a secure element 
        * or cryptoprocessor with storage), we have to allocate a buffer to     
        * hold the generated key material. */                                   
        if( slot->key.data == NULL )                                            
        {                                                                       
            status = psa_allocate_buffer_to_slot( slot, data_length );          
            if( status != PSA_SUCCESS )                                         
                goto exit;                                                      
        }                                                                       
=======                                                                         
>>>>>>> psa: import: Move registered SE support to the driver wrapper

Resolution: In psa_crypto.c, removed all the conflicting code. The check on the
            lifetime added by ac3434fc19 is done naturally when moving the SE
            support to the driver wrapper. In psa_crypto_driver_wrapper.c,
            changed the error code returned in case of unknown location from
            NOT_SUPPORTED to INVALID_ARGUMENT.

===============================================================================

Commit: "tests: psa: Reactivate and expand key management through test driver"

Conflict in test_suite_psa_crypto.data with 8405b5f60d (PR 4012)

<<<<<<< HEAD                                                                    
depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_ECP_DP_CURVE25519_ENABLED
=======                                                                         
depends_on:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_CURVE25519_ENABLED:MBEDTLS_ECDH_C:!MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR
>>>>>>> tests: psa: Reactivate and expand key management through test driver

Resolution:

depends_on:PSA_WANT_ALG_ECDH:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_ECP_DP_CURVE25519_ENABLED:!MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR

===============================================================================

Commit: "tests: psa: Fix export related tests"

Conflict in test_suite_psa_crypto.function with cbcec21684 (PR 3984)
 
<<<<<<< HEAD                                                                    
    exported_size = PSA_EXPORT_KEY_OUTPUT_SIZE( psa_get_key_type( &attributes ),
                                                psa_get_key_bits( &attributes ) );
    ASSERT_ALLOC( exported, exported_size );                                    
                                                                                
=======                                                                         
>>>>>>> tests: psa: Fix export related tests  

Resolution: Remove the conflicting code and change PSA_KEY_EXPORT_MAX_SIZE
            macros introduced by the commit to PSA_EXPORT_KEY_OUTPUT_SIZE
            macros as in cbcec21684.

-------------------------------------------------------------------------------

Conflict in test_suite_psa_crypto.data with 8405b5f60d (PR 4012)

<<<<<<< HEAD                                                                    
depends_on:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C                                
import_export_public_key:"2b7e151628aed2a6abf7158809cf4f3c":PSA_KEY_TYPE_AES:PSA_ALG_CBC_NO_PADDING:0:PSA_ERROR_INVALID_ARGUMENT:""
=======                                                                         
depends_on:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:MBEDTLS_RSA_C                  
import_export_public_key:"2b7e151628aed2a6abf7158809cf4f3c":PSA_KEY_TYPE_AES:PSA_ALG_CBC_NO_PADDING:0:PSA_ERROR_INVALID_ARGUMENT:"2b7e151628aed2a6abf7158809cf4f3c"
>>>>>>> tests: psa: Fix export related tests

Resolution:

depends_on:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C                                
import_export_public_key:"2b7e151628aed2a6abf7158809cf4f3c":PSA_KEY_TYPE_AES:PSA_ALG_CBC_NO_PADDING:0:PSA_ERROR_INVALID_ARGUMENT:"2b7e151628aed2a6abf7158809cf4f3c"

===============================================================================

mpg reviewed (approved):
Re-approving after checking the rebase. Thanks for providing details about the conflicts and how your resolved them, that was very helpful!

stevew817 reviewed (approved):
Instead of going through the force-push diff, I went through the entire PR again to make sure everything still makes sense. Nothing looks out of place to me, still looking good.

ronald-cron-arm commented:
The CI is OK as the only job falling is the ABI-API-checking one which is expected as the internal API psa_driver_wrapper_validate_key() has been removed. Thus ready to be merged.

stevew817 commented:
@ronald-cron-arm I'm integrating this now against our implementation with an opaque crypto driver, and am running into an issue caused by your changes. I know I've approved the PR, too, but I'd like to figure out a way to solve it ASAP.

The issue at hand is how the `import_key` entry point does not have a way to increase the allocated buffer, which is required for opaque drivers (which usually add metadata to a wrapped key for storage). In our integration branch, we've been working under the assumption `driver_wrapper_import_key` would've been directed at a slot instead of a pre-allocated buffer, meaning that we could leverage the static expected_size function to calculate the right size.

Would you be willing to take a patch which converts the signature of `psa_driver_wrapper_import_key` such that its output is a slot instead of a direct buffer/length/outputsize? Like is being done for `psa_driver_wrapper_generate_key` already?

ronald-cron-arm commented:
@stevew817 In development `psa_driver_wrapper_generate_key()` has still a slot as output parameter but not for long: see #3744. I think that the way to go with your issue is to do the same for `psa_import_key()`  as done for `psa_generate_key()` in #3744 regarding the key buffer allocation. Please have a look to #3744 and let me know if this could solve your issue.
